mozjs/js/rust/magic_codegen/src/lib.rs

```diff
@@ -0,0 +1,144 @@
+#![feature(proc_macro)]
+extern crate proc_macro;
+extern crate syn;
+#[macro_use]
+extern crate quote;
+
+use syn::{ Body, VariantData };
+use proc_macro::TokenStream;
+
+#[proc_macro_derive(MagicDom)]
+pub fn magic_dom(input: TokenStream) -> TokenStream {
```

Comment:
In this file in particular, it would help readers (like myself) to have a doc comment on each of these functions that explains what role they play.

```rust
/// The `#[derive(MagicDom)]` implementation.
///
/// Takes a struct spec and creates the magic dom version where the types are
/// stored in an underlying JSObject's slots.
///
/// The struct name must have an underscore, such as `DOMPoint_spec`. Then,
/// the macro will use everything up until the first '_' as the name of the generated
//// magic dom struct. With our earlier example, we'd get `DOMPoint`.
```